### PR TITLE
[WIP] Add more Vitamins

### PR DESCRIPTION
**Pull request by RedSparr0w**
This PR adds more vitamins to the game

Calcium - Adds 1% of base attack extra attack gain per breeding cycle
Carbos - Reduces egg steps required for breeding

A couple of example stats:
_(green lowest → red highest)_
image

The player can still only use a total combined value of 5 per region.

Still has some stuff to be done before it can be merged (`TODO VITAMIN:` in the code)
Even if we decide against the Calcium and Carbos effects, the base work should get done to allow for others.

Notes:
- [ ] Where/when/how should the items be obtainable?
- [ ] Effects adjustment?
- [x] What do we do with the Proteins on peoples current Pokémon?
  - ✔ Add a way to remove vitamins from Pokémon manually
  - ~~Reset and refund?~~
  - ~~Ask if they want to reset them all and refund~~
- [x] Display amount of the vitamins you have used

<details>
<summary>Console stuff</summary>

Generate eff
```js
setVitamin = (vitamin, amt) => {
  App.game.party.caughtPokemon.forEach(p => {
    p.vitaminsUsed[0](0)
    p.vitaminsUsed[1](0)
    p.vitaminsUsed[2](0)
    p.vitaminsUsed[vitamin](amt);
  })
}

output = {};

filteredMons = App.game.party.caughtPokemon.filter(p => [
  'Blissey',
  'Caterpie',
  'Mew',
  'Unown (A)',
  'Mewtwo',
  'Magikarp',
  'Zapdos',
  'Dragonite',
  'Rayquaza',
  'Bulbasaur',
  'Venusaur',
  'Blaziken',
  'Pikachu',
  'Snorlax',
  'Slaking',
  'Gyarados',
].includes(p.name));
filteredMons.forEach(p => {
  p.effortPoints = 0;
  output[p.name] = {};
})
GameHelper.enumStrings(GameConstants.VitaminType).forEach(v => {
  [0, 10, 25, 40].forEach(amt => {
    setVitamin(GameConstants.VitaminType[v], amt);
    filteredMons.forEach(p => {
      output[p.name][v] = output[p.name][v] || {};
      let eff = 0;
      eff = +(SortOptionConfigs[SortOptions.breedingEfficiency].getValue(p).toLocaleString('en-US', { maximumSignificantDigits: 2 }));
      output[p.name][v][amt] = eff;
    })
  });
});

copy(Object.keys(output).map(p => {
    o = `${p}`;
    o += `\t${output[p]['Protein'][0]}`;
    o += `\t${output[p]['Protein'][10]}`;
    o += `\t${output[p]['Protein'][25]}`;
    o += `\t${output[p]['Protein'][40]}`;
    o += `\t${output[p]['Calcium'][10]}`;
    o += `\t${output[p]['Calcium'][25]}`;
    o += `\t${output[p]['Calcium'][40]}`;
    o += `\t${output[p]['Carbos'][10]}`;
    o += `\t${output[p]['Carbos'][25]}`;
    o += `\t${output[p]['Carbos'][40]}`;
    return o;
  }).join('\n'))
```

Get top eff based on vitamin distribution for a mon
```js
getTopEff = (id, totalVitamins = 40) => {
  poke = App.game.party.getPokemon(id)
  
  data = [];
  for(p = 0; p <= 40; p++) {
    poke.vitaminsUsed[0](p)
    for(c = 0; c <= (totalVitamins - p); c++) {
      poke.vitaminsUsed[1](c)
      for(cc = 0; cc <= (totalVitamins - (p + c)); cc++) {
        poke.vitaminsUsed[2](cc)
        data.push([p, c, cc, SortOptionConfigs[SortOptions.breedingEfficiency].getValue(poke)]);
      }
    }
  }
  
  max = Math.max(...data.map(d => d[3]))
  ind = data.findIndex(d => d[3] === max)
  topVal = data[ind];
  return {
    protein: topVal[0],
    calcium: topVal[1],
    carbos: topVal[2],
    eff: topVal[3],
  };
}
```
</details>

### CODE REVIEW

**Comment by Aegyo** [comment]
> What do we do with the Proteins on peoples current Pokémon?

My vote is for allowing refunds for used vitamins at any time. I can't see how it would be hugely abusable, and would allow people to experiment without consequence.

**Comment by RedSparr0w** [comment]
Updated, can now add/remove Vitamins as you like (by right clicking the mon with the vitamin menu open)

**Comment by RedSparr0w** [comment]
image

**Comment by RedSparr0w** [comment]
Adjusted the Carbos effect, so now higher step mons gain a much bigger effect from it than lower step mons:
Steps | Old x 10 | New x 10 | Old x 40 | New x 40
:---: | :---: | :---: | :---: | :---:
4800 | 2727 (56%) | 3230 (67%) | 500 (10%) | 984 (20%)
2400 | 1428 (59%) | 1783 (74%) | 301 (12%) | 731 (30%)
1200 | 748 (62%) | 984 (82%) | 181 (15%) | 543 (45%)
480 | 318 (66%) | 448 (93%) | 92 (19%) | 366 (76%)

File: src/components/pokemonSelectorModal.html [hunk 1]
```diff
@@ -87,21 +87,25 @@ <h5 class="modal-title text-light">Select a Pokémon</h5>&nbsp;
...
+                      <th class="text-center bg-dark align-middle">Vitamins Used</th>
+                      <th class="text-center bg-dark align-middle" data-bind="text: `${GameConstants.VitaminType[VitaminController.currentlySelected()]} Used`">X Used</th>
                       <th class="text-center bg-dark align-middle">Attack Bonus</th>
+                      <th class="text-center bg-dark align-middle">Egg Steps</th>
                     </tr>
                   </thead>
-                  <tbody data-bind="foreach: PartyController.getProteinSortedList()">
-                    <tr class="clickable" data-bind="click: () => $data.useProtein(VitaminController.getMultiplier()), hidden: $data.hideFromProteinList()">
+                  <tbody data-bind="foreach: PartyController.getvitaminSortedList()">
+                    <tr class="clickable" data-bind="click: () => $data.useVitamin(VitaminController.currentlySelected(), VitaminController.getMultiplier()), attr: { onContextMenu: `App.game.party.getPokemonByName('${$data.name}')?.removeVitamin(VitaminController.currentlySelected(), VitaminController.getMultiplier())` }, hidden: $data.hideFromProteinList()">
```
Comment (Jaaslet) [inline src/components/pokemonSelectorModal.html:97]: Can we get the leftclick to stop the input after it's done? For us that have disabled the "block leftclick" setting? I think it's usually just `return false;` at the end

File: src/components/breedingModal.html [hunk 2]
```diff
@@ -356,7 +356,9 @@ <h4><u>Hatchery</u></h4>
 
                     <p class="text-warning"><i>NOTE: Base attack is the value your Pokémon originally has at level 100, before any breeding buffs.</i></p>
 
-                    <p>Any Proteins applied to Pokémon will also give them an additional 1 attack per hatch.</p>
+                    <p>Any Proteins applied to Pokémon will give them an additional 1 attack per hatch, but require 20 more steps.</p>
```
Comment (Jaaslet) [inline src/components/breedingModal.html:359]: The 1 attack per hatch is no longer correct

**Review by Jaaslet — CHANGES_REQUESTED** [review]
Looks good. Some comments:
- There is really no explanation of leftclick removes (as far as i can see). We either need some text, or a toggle button (which basicly switches if it's left or rightclick that removes/adds)
- I think it's fine that they are sold in the league. But i think they should be pushed to later regions. Kanto already contains a lot of new stuff a new player must learn. So i think it would be better if we pushed the choice between vitamins to a later region. Maybe add one in Hoenn and one in Unova or something?
Then we can also inflate the price a bit, so there are more stuff you want to grind money towards
- Both the increased eggsteps and the lower protein might be too much. With Galar level of proteins, there are no mons with a breeding efficiency above 3. Before, the highest was almost 10

File: src/components/breedingModal.html (hunk 2 above)
Comment (RedSparr0w) [reply src/components/breedingModal.html:359]: Sorry, I reverted it back to 1 attack per hatch, but now adds steps.
Refer dev chat for more info.

Comment (Jaaslet) [reply src/components/breedingModal.html:359]: I think you forgot to revert it. At least it looks like that in Egg.ts

File: src/scripts/breeding/Egg.ts [hunk 3]
```diff
@@ -116,8 +124,8 @@ class Egg implements Saveable {
         const gender = PokemonFactory.generateGenderById(pokemonID);
         if (partyPokemon) {
             // Increase attack
-            partyPokemon.attackBonusPercent += Math.max(1, Math.round(GameConstants.BREEDING_ATTACK_BONUS * (efficiency / 100)));
-            partyPokemon.attackBonusAmount += Math.max(0, Math.round(partyPokemon.proteinsUsed() * (efficiency / 100)));
+            partyPokemon.attackBonusPercent += Math.max(1, Math.round((GameConstants.BREEDING_ATTACK_BONUS + partyPokemon.vitaminsUsed[GameConstants.VitaminType.Calcium]()) * (efficiency / 100)));
+            partyPokemon.attackBonusAmount += Math.max(0, Math.round(((partyPokemon.vitaminsUsed[GameConstants.VitaminType.Protein]() / 2) * Math.ceil(partyPokemon.baseAttack / 100)) * (efficiency / 100)));
```
Comment (RedSparr0w) [inline src/scripts/breeding/Egg.ts:128]: Forgot to update this part 😅
```suggestion
            partyPokemon.attackBonusAmount += Math.max(0, Math.round(partyPokemon.vitaminsUsed[GameConstants.VitaminType.Protein]() * (efficiency / 100)));
```

File: src/components/breedingModal.html (hunk 2 above)
Comment (RedSparr0w) [reply src/components/breedingModal.html:359]: Was reverted in the other calculation, but not egg, fixed 👍

File: src/components/pokemonSelectorModal.html (hunk 1 above)
Comment (RedSparr0w) [reply src/components/pokemonSelectorModal.html:97]: I'm not quite sure what you mean, we only have a disable right click option AFAIK.

Left click adds, right click removes _(even with disable right click enabled)_

Comment (Jaaslet) [reply src/components/pokemonSelectorModal.html:97]: I switched left and right in my explanation...

Comment (RedSparr0w) [reply src/components/pokemonSelectorModal.html:97]: I'm still not entirely sure what you mean by "Can we get the right click to stop the input after it's done?"

Comment (PixLSteam) [reply src/components/pokemonSelectorModal.html:97]: Jaaslet wants to prevent the context menu from opening even when the related setting is disabled.

Comment (RedSparr0w) [reply src/components/pokemonSelectorModal.html:97]: Ah I see, thanks, my brain just wasn't clicking 😅

Comment (RedSparr0w) [reply src/components/pokemonSelectorModal.html]: Resolved, also added text stating `Left click to add, right click to remove`

**Comment by RedSparr0w** [comment]
> With Galar level of proteins, there are no mons with a breeding efficiency above 3. Before, the highest was almost 10

Some of the previous top breeders have been nerfed quite a bit by the change, but a ton of the lower efficiency mons have been buffed.

<details>
<summary>New top 20</summary>

<html>
<body>

Pokemon | Eff | Protein | Calcium | Carbos
-- | -- | -- | -- | --
Manaphy | 5.8 | 0 | 11 | 29
Eternamax Eternatus | 5.5 | 0 | 0 | 40
Gyarados | 5 | 0 | 9 | 31
Slaking | 4.4 | 0 | 9 | 31
Tapu Bulu | 4.3 | 0 | 9 | 31
Tapu Lele | 4.2 | 0 | 9 | 31
Phione | 4.1 | 0 | 10 | 30
Totem Wishiwashi | 4.1 | 0 | 10 | 30
Wishiwashi (School) | 4 | 0 | 8 | 32
Tapu Koko | 3.9 | 0 | 11 | 29
Tapu Fini | 3.8 | 0 | 12 | 28
Togekiss | 3.6 | 0 | 11 | 29
Musharna | 3.6 | 0 | 13 | 27
Mega Diancie | 3.6 | 0 | 6 | 34
Clefable | 3.5 | 0 | 12 | 28
Toxicroak | 3.5 | 0 | 11 | 29
Alolan Exeggutor | 3.4 | 0 | 8 | 32
Arcanine | 3.2 | 0 | 9 | 31
Exeggutor | 3.2 | 0 | 9 | 31
Pinkan Exeggutor | 3.2 | 0 | 9 | 31


</body>
</html>
</details>

<details>
<summary>Compare with the old top 20</summary>

<html>
<body>

Pokemon | Eff | Proteins
-- | -- | --
Gyarados | 9.6 | 40
Phione | 8.8 | 40
Manaphy | 8.6 | 40
Magikarp | 8.4 | 40
Morpeko | 6.2 | 40
Morpeko (Hangry) | 6.2 | 40
Totem Togedemaru | 6.1 | 40
Togedemaru | 6 | 40
Cleffa | 5.8 | 40
Igglybuff | 5.8 | 40
Togepi | 5.8 | 40
Spooky Togepi | 5.8 | 40
Surprise Togepi | 5.8 | 40
Togetic | 5.8 | 40
Pachirisu | 5.6 | 40
Clefairy | 5.5 | 40
Pichu | 5.5 | 40
Spiky-eared Pichu | 5.5 | 40
Azurill | 5.5 | 40
Tapu Lele | 5.3 | 40


</body>
</html>
</details>

<details>
<summary>New bottom 20</summary>

<html>
<body>

Dratini | 1.1 | 18 | 0 | 22
-- | -- | -- | -- | --
Larvitar | 1.1 | 17 | 1 | 22
Shedinja | 1.1 | 28 | 0 | 12
Feebas | 1.1 | 27 | 0 | 13
Bagon | 1.1 | 18 | 0 | 22
Gible | 1.1 | 17 | 1 | 22
Axew | 1.1 | 18 | 0 | 22
Deino | 1.1 | 17 | 1 | 22
Goomy | 1.1 | 18 | 0 | 22
Jangmo-o | 1.1 | 18 | 0 | 22
Beldum | 1 | 19 | 0 | 21
Blissey | 0.94 | 13 | 1 | 26
Poipole | 0.94 | 0 | 1 | 39
Dreepy | 0.93 | 21 | 0 | 19
Chansey | 0.9 | 22 | 1 | 17
Happiny | 0.89 | 26 | 0 | 14
Kubfu | 0.79 | 5 | 0 | 35
Meltan | 0.61 | 11 | 0 | 29
Cosmoem | 0.54 | 13 | 0 | 27
Cosmog | 0.44 | 18 | 0 | 22


</body>
</html>
</details>

<details>
<summary>Compare with the old bottom 20</summary>

<html>
<body>

Blacephalon | 0.64 | 40
-- | -- | --
Urshifu (Single Strike) | 0.64 | 40
Urshifu (Rapid Strike) | 0.64 | 40
??? | 0.63 | 40
Deoxys | 0.63 | 40
Nihilego | 0.63 | 40
Pheromosa | 0.63 | 40
Naganadel | 0.61 | 40
Stakataka | 0.61 | 40
Calyrex | 0.61 | 40
Regieleki | 0.6 | 40
Deoxys (Speed) | 0.59 | 40
Deoxys (Attack) | 0.58 | 40
Deoxys (Defense) | 0.58 | 40
Kartana | 0.58 | 40
Poipole | 0.53 | 40
Kubfu | 0.49 | 40
Meltan | 0.44 | 40
Cosmoem | 0.42 | 40
Cosmog | 0.38 | 40


</body>
</html>
</details>

Refer the `+20 per` tab
https://docs.google.com/spreadsheets/d/1N-85Cf0sP9azcg9Ea1ZzepY6euHfYUCyb7IE894hYJI/edit?usp=sharing

**Review by Jaaslet — APPROVED** [review]
Looks good. Thanks for the PR

**Pull request merged**